test/support/test_complex.h

```diff
@@ -0,0 +1,141 @@
+    /* This two types allow us to avoid this situation.                        */                     \
+    using HasDoubleTypeSupportInRuntime = ::std::true_type;                                           \
+    using HasntDoubleTypeSupportInRuntime = ::std::false_type;                                        \
+                                                                                                      \
+    /* long double type generate compile-time error in Kernel code             */                     \
+    /* and we never can use this type inside Kernel                            */                     \
+    using HasntLongDoubleSupportInCompiletime = ::std::false_type;                                    \
+                                                                                                      \
+    TestUtils::run_test_in_kernel(                                                                    \
+        /* labbda for the case when we have support of double type on device */                       \
+        [&]() { run_test<HasDoubleTypeSupportInRuntime, HasntLongDoubleSupportInCompiletime>(); },    \
+        /* labbda for the case when we haven't support of double type on device */                    \
+        [&]() { run_test<HasntDoubleTypeSupportInRuntime, HasntLongDoubleSupportInCompiletime>(); }); \
+                                                                                                      \
+    return TestUtils::done();                                                                         \
+}                                                                                                     \
+                                                                                                      \
+template <typename HasDoubleSupportInRuntime, typename HasLongDoubleSupportInCompiletime>             \
+int                                                                                                   \
+run_test()
+
+// We should use this macros to avoid runtime-error if type double doesn't supported on device.
+// 
+// Example:
+//     template <class T>
+//     void
+//     test(T x, typename std::enable_if<std::is_integral<T>::value>::type* = 0)
+//     {
+//         static_assert((std::is_same<decltype(dpl::conj(x)), dpl::complex<double>>::value), "");
+// 
+//         // HERE IS THE CODE WHICH CALL WE SHOULD AVOID IF DOUBLE IS NOT SUPPORTED ON DEVICE
+//         assert(dpl::conj(x) == dpl::conj(dpl::complex<double>(x, 0)));
+//     }
+//
+//     template <class T>
+//     void test()
+//     {
+//         // ...
+//         test<T>(1);
+//         // ...
+//     }
+//
+//     ONEDPL_TEST_NUM_MAIN
+//     {
+//         // ...
+//         IF_DOUBLE_SUPPORT_IN_RUNTIME(test<int>())
+//         // ...
+//     }
+#define IF_DOUBLE_SUPPORT_IN_RUNTIME(x)                                                               \
```

Comment:
Lines 90-91 actually confuses me.
Macro name tells about RUNTIME, but the implementation tells about COMPILETIME  - `if constexpr`.

Reply:
Fixed